rfcs/2021-04-13-7027-core-extraction.md

```diff
@@ -0,0 +1,319 @@
+carefully curate among our current zoo of feature flags, or other custom build
+goals. It is our belief that by extracting a vector "core" we can:
+
+* reduce iteration costs in engineering work,
+* reduce the burden for casual contributors,
+* enable more focused correctness and performance tests and
+* reduce experimentation costs, pushing toward a fail-fast model of work.
+
+## Motivation
+
+We want to push the vector project into new, more competitive spaces. This will
+mean integrating with more sources of data, allowing our users to do new things
+with their in-flight data and also consistently, release after release, use less
+resources to do it. As noted in the summary build times are long and growing,
+imposing a burden on development. It is _hard_ to benchmark and improve your
+code if the feedback cycle is 30 minutes. Our flat structure raises the barrier
+to entry for contributions; users on discord have referred to our compile times
+as "punishing". If a user contributes a new sink, say, then they are responsible
+for:
+
+* writing the sink and its correctness tests,
+* writing benchmarks,
+* resolving any incompatibilities between their new code and the existing vector
+  model and
+* waiting for CI to approve their changes.
+
+With regard to benchmarks we generally do not request contributors to add them
+as doing so is too complicated. Our flat structure promotes a conflation between
+unit and integrated benchmark tests, not to mention you need a fair bit of
+memory to _link_ vector to run these tests yourself. The feedback loop here is
+hours long. Any incompatibilities between new and existing code are quite
+time-consuming to resolve as our flat structure encourages a "big bang" mindset
+to changes, if not mandates it. The flatness of our project structure does mean,
+however, that relatively simple sources/transforms/sinks are convenient to
+introduce, there is "one place" for such things.
+
+The vector project does have some sub-packages. VRL -- in `lib/vrl` -- is one
+such. The development experience here has a tighter feedback loop than in vector
+proper. VRL benchmarks can be built independent of vector -- avoiding its link
+cost -- and its tests end up clearly focused on VRL itself. This _does_ mean
+that integration tests remain to be written in the top-level of the project but
+the overall experience is improved. Changes in VRL are isolated to VRL, though
+experience with `Lookup`/`LookupBuf` -- see
+[5374](https://github.com/timberio/vector/pull/5374) -- does suggest that
+aggressive integration of a cross-cutting concern will make things challenging.
+
+We hope to capture the benefits to development we've seen in the VRL sub-package
+and make them more broadly available in the vector project.
+
+## The Structure of the Project
+
+Today there are 168 top-level features in the project, as measured on commit
+48d2a84b1b11ba54db7bd892944f2a479238edb4:
+
+```sh
+> cargo read-manifest | jq ".features" | jq 'keys' | wc -l
```

Comment:
Extreme nitpick: this can be `jq ".features | keys | length"` 😄

Reply:
You caught me: I am crap at `jq`. 😂